pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -910,36 +1050,83 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
+
 	stat.ProcessingTime.End(true)
 	stat.UpstreamTime.Start()
 
-	dialer := net.Dialer{
-		Timeout: 2 * time.Second,
-		Control: func(network, address string, c syscall.RawConn) error {
-			var soerr error
-			if err := c.Control(func(su uintptr) {
-				soerr = setSoMark(int(su), ep.GetIdentity())
-			}); err != nil {
-				return err
-			}
-			return soerr
-		},
-	}
-	client.Dialer = &dialer
+	request.Id = dns.Id() // force a random new ID for this request
 
-	conn, err := client.Dial(targetServerAddrStr)
-	if err != nil {
-		err := fmt.Errorf("failed to dial connection to %v: %w", targetServerAddrStr, err)
-		stat.Err = err
-		scopedLog.WithError(err).Error("Failed to dial connection to the upstream DNS server, cannot service DNS request")
-		p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerID, targetServerAddrStr, request, protocol, false, &stat)
-		p.sendRefused(scopedLog, w, request)
-		return
+	// Do not use original source address if the source is known to be in the host networking
+	// namespace
+	useOriginalSrc := !ep.IsHost() && !epAddr.IsLoopback()
+
+	var client *Client
+	var key string
+
+	p.Lock()
+	if useOriginalSrc {
+		// locate client to re-use if possible.
+		key = protocol + "-" + originalSrcAddr.String() + "-" + targetServerAddrStr
+		client = p.DNSClients[key]
+	}
+	if client == nil {
+		dialer := net.Dialer{
+			Timeout: 2 * time.Second,
+			Control: func(network, address string, c syscall.RawConn) error {
+				var soerr error
+
+				if err := c.Control(func(su uintptr) {
+					soerr = setSoMarks(int(su), ipFamily, ep.GetIdentity())
+				}); err != nil {
+					return err
+				}
+				return soerr
+			},
+		}
+		if useOriginalSrc {
+			dialer.LocalAddr = originalSrcAddr
+		}
+
+		client = &Client{
+			Client: dns.Client{
+				Net:            protocol,
+				Dialer:         &dialer,
+				Timeout:        ProxyForwardTimeout,
+				SingleInflight: false,
+			},
+			reqs: make(map[uint16]chan Response),
+		}
+		client.conn, err = client.Dial(targetServerAddrStr)
+		if err != nil {
+			p.Unlock()
+			err := fmt.Errorf("failed to dial connection to %v: %w", targetServerAddrStr, err)
+			stat.Err = err
+			scopedLog.WithError(err).Error("Failed to dial connection to the upstream DNS server, cannot service DNS request")
+			p.NotifyOnDNSMsg(time.Now(), ep, epIPPort, targetServerID, targetServerAddrStr, request, protocol, false, &stat)
+			p.sendRefused(scopedLog, w, request)
+			return
+		}
+		if key != "" {
+			p.DNSClients[key] = client
+		}
 	}
-	defer conn.Close()
+	client.refcount++
+	p.Unlock()
 
-	request.Id = dns.Id() // force a random new ID for this request
-	response, _, err := client.ExchangeWithConn(request, conn)
+	defer func() {
+		p.Lock()
+		client.refcount--
+		if client.refcount == 0 {
+			// Make client unreachable and close it's connection
+			if key != "" {
```

Comment:
nit: I'd prefer `if useOriginalSrc` for consistency

Reply:
Fixed.